### PR TITLE
Add SSL / Auth support

### PR DESCRIPTION
**Pull request by Jotschi**

### CODE REVIEW

File: src/main/java/com/gentics/elasticsearch/client/AbstractElasticsearchClient.java [hunk 1]
```diff
@@ -11,8 +12,19 @@
 	protected int port;
 	protected String hostname;
 	protected String scheme;
+	protected String username;
```
Comment (philippguertler) [inline src/main/java/com/gentics/elasticsearch/client/AbstractElasticsearchClient.java]: Make fields immutable

File: src/main/java/com/gentics/elasticsearch/client/okhttp/ElasticsearchOkClient.java [hunk 2]
```diff
@@ -85,6 +84,34 @@ private OkHttpClient createClient(int connectTimeoutMs, int readTimeoutMs, int w
...
+			// Install the custom trust manager
+			SSLContext sslContext = SSLContext.getInstance("TLS");
+			sslContext.init(null, trustManagers, new java.security.SecureRandom());
+
+			// Create an SSL socket factory with our managers
+			SSLSocketFactory sslSocketFactory = sslContext.getSocketFactory();
+			builder.followRedirects(true);
+			builder.sslSocketFactory(sslSocketFactory, trustManager);
+			if (!isVerifyHostnames()) {
+				builder.hostnameVerifier(new HostnameVerifier() {
```
Comment (philippguertler) [inline src/main/java/com/gentics/elasticsearch/client/okhttp/ElasticsearchOkClient.java]: Check if using lambda is possible

File: src/main/java/com/gentics/elasticsearch/client/okhttp/ElasticsearchOkClient.java [hunk 3]
```diff
@@ -182,6 +209,102 @@ public void onResponse(Call call, Response response) {
...
+			}
+
+			if (converter != null) {
+				client.setConverterFunction(converter);
+			}
+			client.init();
+			return client;
+		}
+
+		public ElasticsearchOkClientBuilder<T> setScheme(String scheme) {
```
Comment (philippguertler) [inline src/main/java/com/gentics/elasticsearch/client/okhttp/ElasticsearchOkClient.java]: javadoc

File: src/main/java/com/gentics/elasticsearch/client/okhttp/ElasticsearchOkClient.java [hunk 4]
```diff
@@ -182,6 +209,102 @@ public void onResponse(Call call, Response response) {
 		});
 	}
 
+	public static class ElasticsearchOkClientBuilder<T> {
```
Comment (philippguertler) [inline src/main/java/com/gentics/elasticsearch/client/okhttp/ElasticsearchOkClient.java]: Name only Builder

File: src/main/java/com/gentics/elasticsearch/client/okhttp/RequestBuilder.java [hunk 5]
```diff
@@ -46,27 +45,35 @@ public RequestBuilder(String method, String path, ElasticsearchOkClient<T> clien
 			body = RequestBody.create(MEDIA_TYPE_NDJSON, builder.toString().getBytes(Charset.defaultCharset()));
 		}
 		this.body = body;
+	}
+
+	public RequestBuilder(String method, String path, ElasticsearchOkClient<T> client, String bulkData) {
 		this.client = client;
 		this.method = method;
+		this.urlBuilder = createUrlBuilder(path);
+		this.body = RequestBody.create(MEDIA_TYPE_NDJSON, bulkData.getBytes(Charset.defaultCharset()));
```
Comment (philippguertler) [inline src/main/java/com/gentics/elasticsearch/client/okhttp/RequestBuilder.java]: Don't use default charset

File: src/main/java/com/gentics/elasticsearch/client/okhttp/RequestBuilder.java [hunk 6]
```diff
@@ -46,27 +45,35 @@ public RequestBuilder(String method, String path, ElasticsearchOkClient<T> clien
 			body = RequestBody.create(MEDIA_TYPE_NDJSON, builder.toString().getBytes(Charset.defaultCharset()));
```
Comment (philippguertler) [inline src/main/java/com/gentics/elasticsearch/client/okhttp/RequestBuilder.java]: Don't use default charset

File: src/main/java/com/gentics/elasticsearch/client/okhttp/RequestBuilder.java [hunk 7]
```diff
@@ -46,27 +45,35 @@ public RequestBuilder(String method, String path, ElasticsearchOkClient<T> clien
...
 			.port(client.getPort())
 			.addPathSegments(path);
 
-		RequestBody body = null;
-		body = RequestBody.create(MEDIA_TYPE_NDJSON, bulkData.getBytes(Charset.defaultCharset()));
-		this.body = body;
-		this.client = client;
-		this.method = method;
+		if (client.hasLogin()) {
+			builder.username(client.getUsername());
```
Comment (philippguertler) [inline src/main/java/com/gentics/elasticsearch/client/okhttp/RequestBuilder.java]: Check if this works or delete if not

File: src/main/java/com/gentics/elasticsearch/client/okhttp/TrustManagerUtil.java [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+		// Password is just needed to create a new keystore. The keystore is only used during runtime.
+		String randomPassword = Long.toString(new Random().nextLong() & Long.MAX_VALUE, 36);
+
+		CertificateFactory certificateFactory = CertificateFactory.getInstance("X.509");
+		Collection<? extends Certificate> certificates = certificateFactory.generateCertificates(certChainStream);
+		if (certificates.isEmpty()) {
+			throw new IllegalArgumentException("expected non-empty set of trusted certificates");
+		}
+
+		// Put the certificates a key store.
```
Comment (philippguertler) [inline src/main/java/com/gentics/elasticsearch/client/okhttp/TrustManagerUtil.java]:
```suggestion
		// Put the certificates in the keystore.
```

File: src/main/java/com/gentics/elasticsearch/client/okhttp/TrustManagerUtil.java [hunk 9]
```diff
@@ -0,0 +1,89 @@
...
+		KeyStore keyStore = KeyStoreUtil.newEmptyKeyStore(randomPassword);
+		int index = 0;
+		for (Certificate certificate : certificates) {
+			String certificateAlias = Integer.toString(index++);
+			keyStore.setCertificateEntry(certificateAlias, certificate);
+		}
+
+		TrustManagerFactory trustManagerFactory = TrustManagerFactory.getInstance(
+			TrustManagerFactory.getDefaultAlgorithm());
```
Comment (philippguertler) [inline src/main/java/com/gentics/elasticsearch/client/okhttp/TrustManagerUtil.java:71]: Check this

**Review by philippguertler — CHANGES_REQUESTED** [review]

File: src/main/java/com/gentics/elasticsearch/client/AbstractElasticsearchClient.java [hunk 10]
```diff
@@ -26,20 +38,87 @@ public AbstractElasticsearchClient(String scheme, String hostname, int port) {
...
 	public String getHostname() {
 		return hostname;
 	}
 
+	@Override
+	public void setLogin(String username, String password) {
+		Objects.requireNonNull(username, "Username must not be null");
+		Objects.requireNonNull(password, "Password must not be null");
+		this.username = username;
+		this.password = password;
```
Comment (deckdom) [inline src/main/java/com/gentics/elasticsearch/client/AbstractElasticsearchClient.java]: How would I clear the login data again?
I'd propose a clearLogin() function for this case

Comment (Jotschi) [reply src/main/java/com/gentics/elasticsearch/client/AbstractElasticsearchClient.java]: You can't. The client is immutable. You would need to create a new one using the builder.

**Pull request merged**